experimental/formatted_string_builder/src/string_builder.rs

```diff
@@ -0,0 +1,259 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::FormattedStringBuilderError;
+
+/** A FormattedStringBuilder with L levels of type annotations. */
+#[cfg_attr(test, derive(Debug, PartialEq))]
+pub struct FormattedStringBuilder<F: Copy, const L: usize> {
+    // This could be Vec<u8> as well, but String makes the encoding more explicit
+    chars: std::string::String,
+    // The first dimension corresponds to the chars (i.e. Chars, code points), the second are the L levels of annotations
+    annotations: Vec<Annotation<F, L>>,
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+enum LocationInPart {
+    Beginning,
+    Inside,
+    End,
+    Single,
+}
+
+// An L-level deep annotation for a single character, using F as field types
+type Annotation<F, const L: usize> = [(LocationInPart, F); L];
```

Comment:
Rust is actually not really ready for this. There are a few assert_eq!(L - 1, L1) runtime check below because const generics cannot be statically evaluated...

Reply:
We'll keep it for now until Rust supports evaluatable generics